### PR TITLE
Add and register gitlab log parsers

### PR DESCRIPTION
**Pull request by alxarch**
## Background

Integrate parsers for Gitlab [logs](https://docs.gitlab.com/ee/administration/logs.html)

## Changes

Adds a new module at `internal/log_analysis/log_processor/parsers/gitlablogs`
- Added parser for Gitlab production_json logs
- Added parser for Gitlab api_json logs
- Added parser for Gitlab integrations_json logs
- Added parser for Gitlab git_json logs
- Added parser for Gitlab exceptions_json logs
- Added parser for Gitlab audit_json logs
- Registered the parsers with internal/log_analysis/log_processor/registry`

## Testing

- How did you test your change?
All parsers have unit tests for the examples provided by gitlab docs

### CODE REVIEW

File: internal/log_analysis/log_processor/parsers/gitlablogs/rails.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+// TypeRails is the type of the GitLabRails log record
+const TypeRails = PantherPrefix + ".Rails"
+
+// RailsDesc describes the GitLabRails log record
+var RailsDesc = `GitLab log for Rails controller requests received from GitLab
+Reference: https://docs.gitlab.com/ee/administration/logs.html#production_jsonlog`
+
+// Rails is a a GitLab Rails controller log line from a non-API endpoint
+// TODO: Check more samples from [Lograge](https://github.com/roidrage/lograge/) JSON output to find missing fields
+type Rails struct {
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/gitlablogs/rails.go:40]: Should be "ProductionJSON"?

File: internal/log_analysis/log_processor/parsers/gitlablogs/api.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+// API is a a GitLab log line from an internal API endpoint
+// TODO: Check more samples from [Lograge](https://github.com/roidrage/lograge/) JSON output to find missing fields
+type API struct {
+	Time           *time.Time   `json:"time" validate:"required" description:"The request timestamp"`
+	Severity       *string      `json:"severity" validate:"required" description:"The log level"`
+	Duration       *float64     `json:"duration" validate:"required" description:"The time spent serving the request (in milliseconds)"`
+	DB             *float64     `json:"db,omitempty" description:"The time spent quering the database (in milliseconds)"`
+	View           *float64     `json:"view,omitempty" description:"The time spent rendering the view for the Rails controller (in milliseconds)"`
+	Status         *int         `json:"status" validate:"required" description:"The HTTP response status code"`
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/gitlablogs/api.go:46]: Should be `*int16`. It makes little difference for the Go code, however we use this type when creating the Glue Data Catalog where this data will be stored. 
`int16` gets [mapped](https://github.com/panther-labs/panther/blob/89b6dd7cfdfc97afa9007ecac1738057b3f37841/tools/cfngen/gluecf/schema.go#L216-L217) to `smallint`, while `int` will gets mapped to `bigint`. When we eventually convert the data to parquet data format, we could have significant compression benefits when going with `smallint`.

File: internal/log_analysis/log_processor/parsers/gitlablogs/rails.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	Method             *string      `json:"method" validate:"required" description:"The HTTP method of the request"`
+	Path               *string      `json:"path" validate:"required" description:"The URL path for the request"`
+	Format             *string      `json:"format" validate:"required" description:"The response output format"`
+	Controller         *string      `json:"controller" validate:"required" description:"The Rails controller class name"`
+	Action             *string      `json:"action" validate:"required" description:"The Rails controller action"`
+	Status             *int         `json:"status" validate:"required" description:"The HTTP response status code"`
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/gitlablogs/rails.go:46]: See comment above: Better go for `*int16`

File: internal/log_analysis/log_processor/parsers/gitlablogs/api.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+const TypeAPI = PantherPrefix + ".API"
+
+// APIDesc describes the GitLabAPI log record
+var APIDesc = `GitLab log for API requests received from GitLab
+Reference: https://docs.gitlab.com/ee/administration/logs.html#api_jsonlog`
+
+// API is a a GitLab log line from an internal API endpoint
+// TODO: Check more samples from [Lograge](https://github.com/roidrage/lograge/) JSON output to find missing fields
+type API struct {
+	Time           *time.Time   `json:"time" validate:"required" description:"The request timestamp"`
```
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/log_processor/parsers/gitlablogs/api.go:41]: >     * I see a file included in this PR `internal/log_analysis/log_processor/parsers/gitlablogs/debug.test`. Leftover?
Oops, sorry about the file. It's from debugging. I usually have it on my .gitignore rules by default. Fixing the PR

>     * Try to deploy the code to your AWS account and verify that the necessary tables get created in Glue Data Catalog. You can see the tables when going to `https://eu-west-1.console.aws.amazon.com/glue/home?region=eu-west-1#` (you might need to change the region depending where you have it deployed)
Will do that, if i got this right the tables are created for all registered parsers

File: internal/log_analysis/log_processor/parsers/gitlablogs/api.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+	jsoniter "github.com/json-iterator/go"
+	"go.uber.org/zap"
+
+	"github.com/panther-labs/panther/internal/log_analysis/log_processor/parsers"
+	"github.com/panther-labs/panther/internal/log_analysis/log_processor/parsers/timestamp"
+)
+
+// TypeAPI is the type of the GitLabAPI log record
+const TypeAPI = PantherPrefix + ".API"
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/gitlablogs/api.go:32]: I am thinking probably better to go with `APIJson` or `API.JSON` (@rleighton  any preference?)
This looks more consistent with GitLab's naming as well and will save us from future trouble in case of there parsers if we decide we need to support both `application.log` and `application_json.log`.

File: internal/log_analysis/log_processor/parsers/gitlablogs/rails.go (hunk 1 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/parsers/gitlablogs/rails.go:40]: It's not really clear from the docs. I got the impression that the file changes names depending on the `env` that it's deployed (ie `production`, `development` etc). From what I understood it's just the logging of a popular(?) Rails logger called `Lograge` for the controller endpoints.

File: internal/log_analysis/log_processor/parsers/gitlablogs/exceptions.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+const TypeExceptions = PantherPrefix + ".Exceptions"
+
+// ExceptionsDesc describes the Git log record
+var ExceptionsDesc = `GitLab log file containing changes to group or project settings 
+Reference: https://docs.gitlab.com/ee/administration/logs.html#audit_jsonlog`
+
+// Exceptions is a a GitLab log line from a failed interaction with git
+type Exceptions struct {
+	Severity           *string      `json:"severity" validate:"required" description:"The log level"`
+	Time               *time.Time   `json:"time" validate:"required" description:"The event timestamp"`
```
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/log_processor/parsers/gitlablogs/exceptions.go:41]: I think for exceptions the `exception.message/class` would be required as well. Will change that.

File: internal/log_analysis/log_processor/parsers/gitlablogs/api.go (hunk 5 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/parsers/gitlablogs/api.go:32]: The text formats are horrific multi-line rails literature, probably unusable to a parser. Other than that this is why I put the type in a const at the top of every file so names are easy to change

File: internal/log_analysis/log_processor/parsers/gitlablogs/api.go (hunk 2 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/parsers/gitlablogs/api.go:46]: does this apply to `floats` as well?

Comment (kostaspap) [reply internal/log_analysis/log_processor/parsers/gitlablogs/api.go:46]: > does this apply to `floats` as well?

Yes. If we're sure that a value will be float32 better choose that over float64. If we are not 100% sure, better stick with float64. Column type is not something that we can change easily once we picked it.

File: internal/log_analysis/log_processor/parsers/gitlablogs/api.go (hunk 5 above)
Comment (kostaspap) [reply internal/log_analysis/log_processor/parsers/gitlablogs/api.go:30]: Fair enough. We can stick with current naming.

File: internal/log_analysis/log_processor/parsers/gitlablogs/api_test.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  "user_id":1,
+  "username":"root",
+  "queue_duration":100.31,
+  "gitaly_calls":30,
+  "gitaly_duration":5.36
+}`
+
+	expectedTime := time.Date(2018, 10, 29, 12, 49, 42, int(123*time.Millisecond), time.UTC)
+	strTime := expectedTime.Format(time.RFC3339Nano)
+	_ = strTime
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/gitlablogs/api_test.go:57]: Not sure, why is this required? Same goes for all tests that have similar code.

File: internal/log_analysis/log_processor/parsers/gitlablogs/integrations.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/panther-labs/panther/internal/log_analysis/log_processor/parsers"
+	"github.com/panther-labs/panther/internal/log_analysis/log_processor/parsers/timestamp"
+)
+
+// TypeIntegrations is the log type of GitLabIntegrations
+const TypeIntegrations = PantherPrefix + ".Integrations"
+
+// IntegrationsDesc describes the GitLabIntegrations log record
+var IntegrationsDesc = `GitLab log with information about integrations activities such as Jira, Asana, and Irker services.
+Reference: https://docs.gitlab.com/ee/administration/logs.html#api_jsonlog`
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/gitlablogs/integrations.go:34]: Link seems wrong.

File: internal/log_analysis/log_processor/parsers/gitlablogs/api_test.go (hunk 7 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/parsers/gitlablogs/api_test.go:57]: Helps with the debugger to be able to watch the value. All this will be easier if errors are returned from parse, will remove

File: internal/log_analysis/log_processor/parsers/gitlablogs/integrations.go (hunk 8 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/parsers/gitlablogs/integrations.go]: Fixed

File: internal/log_analysis/log_processor/parsers/gitlablogs/api_test.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	checkGitLabAPI(t, log, expectedEvent)
+}
+func TestGitLabAPIType(t *testing.T) {
+	parser := &APIParser{}
+	require.Equal(t, "GitLab.API", parser.LogType())
+}
+
+func checkGitLabAPI(t *testing.T, log string, expectedEvent *API) {
+	expectedEvent.SetEvent(expectedEvent)
+	parser := &APIParser{}
```
[the remark this answers is not in the document]
Comment (alxarch) [reply internal/log_analysis/log_processor/parsers/gitlablogs/api_test.go]: Fixed

File: internal/log_analysis/log_processor/parsers/gitlablogs/git_test.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	log := `{
+		"severity":"ERROR",
+		"time":"2019-07-19T22:16:12.528Z",
+		"correlation_id":"FeGxww5Hj64",
+		"message":"Command failed [1]: /usr/bin/git --git-dir=/Users/vsizov/gitlab-development-kit/gitlab/tmp/tests/gitlab-satellites/group184/gitlabhq/.git --work-tree=/Users/vsizov/gitlab-development-kit/gitlab/tmp/tests/gitlab-satellites/group184/gitlabhq merge --no-ff -mMerge branch 'feature_conflict' into 'feature' source/feature_conflict\n\nerror: failed to push some refs to '/Users/vsizov/gitlab-development-kit/repositories/gitlabhq/gitlab_git.git'"
+	}`
+
+	expectedTime := time.Date(2019, 7, 19, 22, 16, 12, int(528*time.Millisecond), time.UTC)
+	strTime := expectedTime.Format(time.RFC3339Nano)
+	_ = strTime
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/gitlablogs/git_test.go:43]: Can you remove as well?

File: internal/log_analysis/log_processor/parsers/gitlablogs/integrations_test.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  "service_class":"JiraService",
+  "project_id":3,
+  "project_path":"namespace2/project2",
+  "message":"Successfully posted",
+  "client_url":"http://jira.example.com"
+}`
+
+	expectedTime := time.Date(2018, 9, 6, 17, 15, 16, int(365*time.Millisecond), time.UTC)
+	strTime := expectedTime.Format(time.RFC3339Nano)
+	_ = strTime
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/gitlablogs/integrations_test.go:74]: remove

File: internal/log_analysis/log_processor/parsers/gitlablogs/rails_test.go [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    "lib/gitlab/session.rb:11:in 'with_session'",
+    "app/controllers/application_controller.rb:450:in 'set_session_storage'",
+    "app/controllers/application_controller.rb:444:in 'set_locale'",
+    "ee/lib/gitlab/jira/middleware.rb:19:in 'call'"
+  ]
+}`
+
+	expectedTime := time.Date(2019, 11, 14, 13, 12, 46, int(156*time.Millisecond), time.UTC)
+	strTime := expectedTime.Format(time.RFC3339Nano)
+	_ = strTime
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/gitlablogs/rails_test.go:119]: Remove
